### DISCUSSION TITLE
Custom Endpoint Include Functions Library

### DISCUSSION THREAD

**Discussion by Thermal-Shock** [directus/directus#12086] [Archive]
I'm likely writing a bunch of custom endpoints, but I would them to import/require from a set of functions in another javascript file. 

How would I import them and potentially add them to my package.json so that when I do a `directus-extension build` that they are included in the bundle or in the correct place to be imported? 

Clearly `directus:extension` does not allow a comma-separated list in `source `and I've also tried various require/import statements, but I cannot seem to find the combination that puts everything in the right place. 

Does anyone have an example of how to do this?

Is the only option to include the functions within the the endpoint index.js itself?

All the best.

**Comment by rijkvanzanten** [comment]
Heya! You should be able to rely on regular node CJS `require` calls to import other files / libraries:

```js
const myUtil = require("./utils/example");

module.exports = function registerEndpoints(router) {
  router.get('/example', (req, res) => {
    const result = myUtil();
    res.send(result);
  });
}
```

**Reply by Thermal-Shock** [reply]
Perfect. Just what I was looking for. Thanks so much. All the best.

**Reply by rokdd** [reply]
Little side note: When you watch the extension it does not reload the "external" file and also in my case the build command does not copy the file in the dist folder.

**Reply by Thermal-Shock** [reply]
Thanks for letting me know @rokdd. What additional steps need to be taken for a complete solution @rijkvanzanten?

**Reply by rijkvanzanten** [reply]
You could use something like `nodemon` to auto restart the api whenever you make a change to files you want to respond to. As for the core reloading/extensions-sdk, we're looking at various improvements on that front, so that's TBD

**Reply by rokdd** [reply]
@Thermal-Shock @rijkvanzanten : The reloading is for me not such a big issue (the autoreload is nice but already more as minimal necessary). I see more a problem, that the files are not copied properly to dist. I am new to nodejs, but I guess the ["files" directive](https://docs.npmjs.com/cli/v8/configuring-npm/package-json) could make sense to use in the package.json of the extension?

**Reply by rijkvanzanten** [reply]
When you say "files aren't copied to dist", do you mean when using the extensions sdk? The API extensions don't require any build tool or anything, so I'm not entirely sure what the thing is that is doing the copying atm

**Reply by nickrum** [reply]
When using the Extensions SDK to develop API extensions, the SDK will also bundle API extensions into a single entrypoint for ease of installing them. This only works when using ESM import though:

```js
import myUtil from "./utils/example.js";

export default (router) => {
  router.get('/example', (req, res) => {
    const result = myUtil();
    res.send(result);
  });
};
```

**Reply by rijkvanzanten** [reply]
Ahh that might be the confusion here then

**Reply by rokdd** [reply]
Great guys!! I read that I should do a JS import but now I figured it out:
```
//index.js
import { y as x_lib} from './lib/xyu.js'

export default (router, { services, exceptions, logger,env,database }) => {
	const x_service  = x_lib(services,logger)

//xyu.js
module.exports = {
    y: function (services, logger) {
```
